pkg/scheduler/plugins/coescheduling/core/gang.go

```diff
@@ -0,0 +1,436 @@
+
+	mode := pg.Annotations[extension.AnnotationGangMode]
+	if mode != extension.GangModeStrict && mode != extension.GangModeNonStrict {
+		klog.Errorf("podGroup's annotation GangModeAnnotation illegal, gangName: %v, value: %v",
+			gang.Name, pg.Annotations[extension.AnnotationGangMode])
+		mode = extension.GangModeStrict
+	}
+	gang.Mode = mode
+
+	// here we assume that coescheduling's CreateTime equal with the podGroup CRD CreateTime
+	gang.CreateTime = pg.CreationTimestamp.Time
+
+	waitTime, err := ParsePgTimeoutSeconds(*pg.Spec.ScheduleTimeoutSeconds)
+	if err != nil {
+		klog.Errorf("podGroup's ScheduleTimeoutSeconds illegal, gangName: %v, value: %v",
+			gang.Name, pg.Spec.ScheduleTimeoutSeconds)
+		if args.DefaultTimeoutSeconds != nil {
+			waitTime = args.DefaultTimeoutSeconds.Duration
+		}
+	}
+	gang.WaitTime = waitTime
+
+	groupSlice, err := StringToGangGroupSlice(pg.Annotations[extension.AnnotationGangGroups])
+	if err != nil {
+		klog.Errorf("podGroup's annotation GangGroupsAnnotation illegal, gangName: %v, value: %v",
+			gang.Name, pg.Annotations[extension.AnnotationGangGroups])
+	}
+	gang.GangGroup = groupSlice
+
+	gang.GangFrom = GangFromPodGroupCrd
+
+	gang.HasGangInit = true
+
+	klog.Infof("TryInitByPodGroup done, gangName: %v, minRequiredNumber: %v, totalChildrenNum: %v, "+
+		"mode: %v, waitTime: %v, groupSlice: %v", gang.Name, gang.MinRequiredNumber, gang.TotalChildrenNum,
+		gang.Mode, gang.WaitTime, gang.GangGroup)
+}
+
+func (gang *Gang) deletePod(pod *v1.Pod) bool {
+	if pod == nil {
+		return false
+	}
+
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	podId := GetNamespaceSplicingName(pod.Namespace, pod.Name)
+	klog.Infof("Delete pod from gang: %v, podName: %v", gang.Name, podId)
+
+	delete(gang.Children, podId)
+	delete(gang.WaitingForBindChildren, podId)
+	delete(gang.BoundChildren, podId)
+	delete(gang.ChildrenScheduleRoundMap, podId)
+	if len(gang.BoundChildren)+len(gang.WaitingForBindChildren) == 0 {
+		gang.ResourceSatisfied = false
+	}
+	if gang.GangFrom == GangFromPodAnnotation {
+		if len(gang.Children) == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func (gang *Gang) getGangWaitTime() time.Duration {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	return gang.WaitTime
+}
+
+func (gang *Gang) getChildrenNum() int {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	return len(gang.Children)
+}
+
+func (gang *Gang) getGangMinNum() int {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	return gang.MinRequiredNumber
+}
+
+func (gang *Gang) getGangTotalNum() int {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	return gang.TotalChildrenNum
+}
+
+func (gang *Gang) getGangMode() string {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	return gang.Mode
+}
+
+func (gang *Gang) getGangAssumedPods() int {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	return len(gang.WaitingForBindChildren) + len(gang.BoundChildren)
+}
+
+func (gang *Gang) getGangScheduleCycle() int {
```

Comment:
getGangScheduleCycle-> getScheduleCycle; the naming rule not equal: "setScheduleCycleValid" has no "gang"